hyperdrive.go

```diff
@@ -68,31 +79,71 @@ func (hyperdrive *hyperdrive) AcceptTick(t time.Time) {
 		ticks++
 		hyperdrive.ticksPerShard[shardHash] = ticks
 
-		if hyperdrive.ticksPerShard[shardHash] > hyperdrive.timeoutThreshold {
-			// 2. After a number of ticks send a TimedOut transition to the shard
+		if ticks > NumTicksToTriggerTimeOut {
+			// 2. Send a TimedOut transition to the shard
+			if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+				replica.Transition(consensus.TimedOut{Time: t})
+			}
 		}
 	}
 }
 
-func (hyperdrive *hyperdrive) AcceptPropose(shardHash sig.Hash, proposed block.Block) {
-	// TODO:
+func (hyperdrive *hyperdrive) AcceptPropose(shardHash sig.Hash, proposed block.SignedBlock) {
 	// 1. Verify the block is well-formed
-	// 2. Verify the signatory of the block
+	if !hyperdrive.validateBlock(shardHash, proposed) {
+		return
+	}
+
+	if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+		replica.Transition(consensus.Proposed{SignedBlock: proposed})
+	}
 }
 
 func (hyperdrive *hyperdrive) AcceptPreVote(shardHash sig.Hash, preVote block.SignedPreVote) {
-	// TODO:
 	// 1. Verify the pre-vote is well-formed
+	if preVote.String() == (block.SignedPreVote{}).String() {
+		return
+	}
+	if preVote.PreVote.Block != nil {
+		if !hyperdrive.validateBlock(shardHash, *preVote.PreVote.Block) {
+			return
+		}
+	}
+	if preVote.PreVote.Round < 0 || preVote.PreVote.Height < 0 {
+		return
+	}
+
 	// 2. Verify the signatory of the pre-vote
+	if !hyperdrive.verifySignature(shardHash, []byte(preVote.PreVote.String()), preVote.Signature, preVote.Signatory) {
+		return
+	}
+
+	if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+		replica.Transition(consensus.PreVoted{SignedPreVote: preVote})
+	}
 }
 
 func (hyperdrive *hyperdrive) AcceptPreCommit(shardHash sig.Hash, preCommit block.SignedPreCommit) {
-	// TODO:
 	// 1. Verify the pre-commit is well-formed
+	if preCommit.String() == (block.SignedPreCommit{}).String() {
+		return
+	}
+	if !hyperdrive.validatePolka(shardHash, preCommit.PreCommit.Polka) {
+		return
+	}
+
 	// 2. Verify the signatory of the pre-commit
+	if !hyperdrive.verifySignature(shardHash, []byte(preCommit.PreCommit.String()), preCommit.Signature, preCommit.Signatory) {
+		return
+	}
+
+	if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+		replica.Transition(consensus.PreCommitted{SignedPreCommit: preCommit})
+	}
 }
 
 func (hyperdrive *hyperdrive) AcceptShard(shard shard.Shard, blockchain block.Blockchain) {
+	// TODO: Will there be a scenario where a replica is already present for the shard?
```

Comment:
This is a great point. We should check to see if there is already a Replica associated with this Shard. If there is, we short-circuit and do nothing.

Reply:
Added. 👍